core/trino-main/src/main/java/io/trino/operator/unnest/UnnestBlockBuilder.java

```diff
@@ -13,233 +13,131 @@
-        if (usingCopiedBlock) {
-            outputBlock = outputBlockBuilder.build();
-        }
-        else {
-            outputBlock = new DictionaryBlock(positionCount, source, ids);
+        if (nullFinder.getNullElementIndex() == NULL_NOT_FOUND) {
+            source = source.appendNull();
+            nullFinder.setNullElementIndex(source.getPositionCount() - 1);
         }
 
-        // Flush stored state, so that ids cannot be modified after the dictionary has been constructed
-        clearCurrentOutput();
-
-        return outputBlock;
+        return buildWithNullsByDictionary(
+                outputEntries,
+                startPosition,
+                inputBatchSize,
+                outputPositionCount,
+                nullFinder.getNullElementIndex(),
+                lengths);
     }
 
-    /**
-     * Append element at position {@code index} in the source block to output block
-     */
-    public void appendElement(int index)
+    private Block buildWithNullsByDictionary(
+            int[] requiredOutputEntries,
+            int offset,
+            int inputBatchSize,
+            int outputPositionCount,
+            int nullIndex,
+            int[] lengths)
     {
-        checkState(source != null, "source is null");
-        checkElementIndex(index, source.getPositionCount());
+        verify(nullIndex != NULL_NOT_FOUND, "nullIndex is -1");
 
-        if (usingCopiedBlock) {
-            type.appendTo(source, index, outputBlockBuilder);
-        }
-        else {
-            appendId(index);
-        }
-    }
+        int position = 0;
+        int[] ids = new int[outputPositionCount];
 
-    /**
-     * Append range of elements starting at {@code startPosition} and length {@code length}
-     * from the source block to output block.
-     * <p>
-     * Purpose of this method is to avoid repeated range checks for every invocation of {@link #appendElement}.
-     */
-    public void appendRange(int startPosition, int length)
-    {
-        // check range validity
-        checkState(source != null, "source is null");
-        checkPositionIndexes(startPosition, startPosition + length, source.getPositionCount());
-
-        if (usingCopiedBlock) {
-            for (int i = 0; i < length; i++) {
-                type.appendTo(source, startPosition + i, outputBlockBuilder);
-            }
-        }
-        else {
-            for (int i = 0; i < length; i++) {
-                appendId(startPosition + i);
-            }
-        }
-    }
+        for (int i = 0; i < inputBatchSize; i++) {
+            int entryCount = lengths[offset + i];
 
-    public void appendNull()
-    {
-        if (usingCopiedBlock) {
-            outputBlockBuilder.appendNull();
-        }
-        else {
-            int nullIndex = nullFinder.getNullElementIndex();
-            if (nullIndex == -1) {
-                startUsingCopiedBlock();
-                appendNull();
+            if (nullCheckBlock == null) {
```

Comment:
when do we expect `nullCheckBlock` to be null if we get a call to `buildWithNulls`?
Can we short-circuit `buildWithNulls` to `buildWithoutNulls` if `nullCheckBlock` is not set?

Reply:
buildWithNulls invocation means that we need to append nulls in the output, irrespective of whether the input block has nulls or not. so we won't be able to short circuit